sklearn/base.py

```diff
@@ -566,6 +574,28 @@ def fit_transform(self, X, y=None, **fit_params):
             # fit method of arity 2 (supervised transformation)
             return self.fit(X, y, **fit_params).transform(X)
 
+    @property
+    def n_features_out_(self):
+        return self._n_features_out
+
+    @n_features_out_.setter
+    def n_features_out_(self, val):
+        self._n_features_out = val
+
+
+class ComponentsMixin:
+    @property
+    def n_features_out_(self):
+        if hasattr(self, 'n_components_'):
+            # n_components could be auto or None
+            # this is more likely to be an int
+            n_features = self.n_components_
+        elif hasattr(self, 'n_components') and self.n_components is not None:
+            n_features = self.n_components
+        elif hasattr(self, 'components_'):
+            n_features = self.components_.shape[0]
+        return n_features
```

Comment:
should there be a "default" value here? Maybe `None`?

Reply:
Here? This is for ``ComponentsMixin``. If it doesn't have ``components`` it probably shouldn't have the ``ComponentsMixin``.

In general: the route I'm taking here is to require the user to set it to ``None`` and error otherwise in the tests. That allowed me to test that it's actually implemented.
What we want to enforce for third-party estimators is a slightly different discussion. This PR currently adds new requirements in ``check_estimator``.